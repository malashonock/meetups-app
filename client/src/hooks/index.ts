export * from './useMeetupQuery';
export * from './useNewsArticleQuery';
export * from './useStaticFileQuery';
export * from './useAuthStore';
