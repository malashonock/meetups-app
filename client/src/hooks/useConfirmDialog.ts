import { useContext } from 'react';

import { ConfirmDialogContext, ConfirmFn } from 'components';
import { Optional } from 'types';

export const useConfirmDialog = (): ConfirmFn => {
  return useContext(ConfirmDialogContext);
};
