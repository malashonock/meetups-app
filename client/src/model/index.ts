export * from './meetup.model';
export * from './news.model';
export * from './user.model';
export * from './login.model';
