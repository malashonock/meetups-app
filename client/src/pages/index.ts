export * from './LoginPage/LoginPage';
export * from './MeetupsPage/MeetupsPage';
export * from './ViewMeetupPage/ViewMeetupPage';
export * from './CreateMeetupPage/CreateMeetupPage';
export * from './EditMeetupPage/EditMeetupPage';
export * from './NewsPage/NewsPage';
export * from './ViewNewsPage/ViewNewsPage';
export * from './CreateNewsPage/CreateNewsPage';
export * from './EditNewsPage/EditNewsPage';
export * from './NotFoundPage/NotFoundPage';
