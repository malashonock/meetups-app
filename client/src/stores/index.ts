export * from './root.store';
export * from './auth.store';
export * from './user.store';
export * from './ui.store';
export * from './meetup.store';
export * from './news.store';
