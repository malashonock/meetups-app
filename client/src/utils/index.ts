export * from './string';
export * from './date';
export * from './number';
export * from './file';
export * from './array';
export * from './test';
export * from './dom';
