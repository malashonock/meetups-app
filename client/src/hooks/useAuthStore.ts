import { useContext } from 'react';

import { AuthStore } from 'stores';
import { RootContext } from 'components';

export const useAuthStore = (): AuthStore => {
  return useContext(RootContext).authStore;
};
