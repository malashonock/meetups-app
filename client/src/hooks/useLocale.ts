import { useContext } from 'react';

import { UiStore, Locale } from 'stores';
import { RootContext } from 'components';
import { Optional } from 'types';

export type UseLocaleResult = [
  locale: Optional<Locale>,
  setLocale: (locale: Locale) => void,
];

export const useLocale = (): UseLocaleResult => {
  const uiStore: Optional<UiStore> = useContext(RootContext)?.uiStore;

  const locale = uiStore?.locale;

  const setLocale = (locale: Locale): void => {
    if (uiStore) {
      uiStore.locale = locale;
    }
  };

  return [locale, setLocale];
};
