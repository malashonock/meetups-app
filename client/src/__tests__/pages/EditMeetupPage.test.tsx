/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-wait-for-side-effects */

import { PropsWithChildren } from 'react';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { EditMeetupPage } from 'pages';
import { MeetupFields } from 'model';
import {
  mockImageWithUrl2,
  mockMeetup,
  mockUser,
  mockUser2,
} from 'model/__fakes__';
import { dropFile } from 'utils';
import { useLocale, useMeetup, useUserStore } from 'hooks';
import { Locale, Meetup, RootStore } from 'stores';

jest.setTimeout(15 * 1000);

jest.mock('utils/file');

// Mock hooks
jest.mock('hooks', () => {
  return {
    ...jest.requireActual('hooks'),
    useUserStore: jest.fn(),
    useMeetup: jest.fn(),
    useLocale: jest.fn(),
  };
});
const mockUseUserStore = useUserStore as jest.MockedFunction<
  typeof useUserStore
>;
const mockUseMeetup = useMeetup as jest.MockedFunction<typeof useMeetup>;
const mockUseLocale = useLocale as jest.MockedFunction<typeof useLocale>;

const mockMeetupUpdate = jest.spyOn(Meetup.prototype, 'update');

beforeEach(() => {
  const { authStore, meetupStore } = new RootStore();
  const { userStore } = authStore;
  userStore.users = [mockUser, mockUser2];
  mockUseUserStore.mockReturnValue(userStore);

  const mockInitializedMeetup = new Meetup(mockMeetup, meetupStore);
  mockInitializedMeetup.image = mockMeetup.image;
  mockInitializedMeetup.isInitialized = true;
  mockUseMeetup.mockReturnValue(mockInitializedMeetup);

  mockUseLocale.mockReturnValue([Locale.RU, jest.fn()]);
});

afterEach(() => {
  jest.resetAllMocks();
});

const MockRouter = ({ children }: PropsWithChildren): JSX.Element => (
  <MemoryRouter initialEntries={['/meetups', '/meetups/aaa/edit']}>
    <Routes>
      <Route path="/meetups">
        <Route index element={<h1>Meetups page</h1>} />
        <Route path=":id">
          <Route index element={<h1>View meetup</h1>} />
          <Route path="edit" element={children} />
        </Route>
      </Route>
    </Routes>
  </MemoryRouter>
);

const getSubjectInput = () =>
  screen.getByLabelText('formFields.meetup.topic.label');
const getExcerptInput = () =>
  screen.getByLabelText('formFields.meetup.description.label');
const getSpeakerSelect = () =>
  screen.getByTestId('select-field').querySelector('input') as HTMLInputElement;
const getStartDatePicker = () =>
  screen.getByLabelText(
    'formFields.meetup.datetimeStart.label',
  ) as HTMLInputElement;
const getFinishDatePicker = () =>
  screen.getByLabelText(
    'formFields.meetup.datetimeFinish.label',
  ) as HTMLInputElement;
const getLocationInput = () =>
  screen.getByLabelText('formFields.meetup.location.label');
const getImageDropbox = () => screen.getByTestId('image-dropbox');
const getImagePreview = () => screen.getByTestId('image-preview');
const getImage = () =>
  screen.getByAltText('imagePreview.imgAlt') as HTMLImageElement;
const getSubmitBtn = () => screen.getByText('formButtons.save');
const getCancelBtn = () => screen.getByText('formButtons.cancel');
const getPreviewBtn = () => screen.getByText('formButtons.preview');
const getClearImageBtn = () => screen.getByTestId('clear-button');

const mockUpdatedMeetup: MeetupFields = {
  subject: 'Updated meetup topic',
  excerpt: 'Updated meetup description',
  author: mockUser,
  speakers: [mockUser, mockUser2],
  place: 'room 321',
  start: new Date(2023, 2, 20, 15, 0),
  finish: new Date(2023, 2, 20, 16, 30),
  image: mockImageWithUrl2,
};

const editFields = async () => {
  // Edit subject
  await act(() => {
    userEvent.clear(getSubjectInput());
    userEvent.type(getSubjectInput(), mockUpdatedMeetup.subject);
  });

  // Edit excerpt
  await act(() => {
    userEvent.clear(getExcerptInput());
    userEvent.type(getExcerptInput(), mockUpdatedMeetup.excerpt);
  });

  // Select another speaker
  await act(() => {
    userEvent.click(getSpeakerSelect());
  });

  await act(async () => {
    userEvent.click(await screen.findByText(mockUser2.fullName));
  });

  // Select another start date/time
  await waitFor(() => {
    userEvent.clear(getStartDatePicker());
    userEvent.type(getStartDatePicker(), '20 Mar 2023 15:00');
  });

  // Select another finish date/time
  await waitFor(() => {
    userEvent.clear(getFinishDatePicker());
    userEvent.type(getFinishDatePicker(), '20 Mar 2023 16:30');
  });

  // Edit location
  await act(() => {
    userEvent.clear(getLocationInput());
    userEvent.type(getLocationInput(), mockUpdatedMeetup.place || '');
  });

  // Upload another image
  await act(() => {
    userEvent.click(getClearImageBtn());
  });

  await act(() => {
    dropFile(getImageDropbox(), mockImageWithUrl2);
  });
};

describe('EditMeetupPage', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<EditMeetupPage />, { wrapper: MockRouter });

    expect(asFragment()).toMatchSnapshot();
  });

  it('on open, should pre-fill form inputs with the current field values', () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    expect(getSubjectInput()).toHaveValue(mockMeetup.subject);
    expect(getExcerptInput()).toHaveValue(mockMeetup.excerpt);
    expect(screen.getByText(mockMeetup.author!.fullName)).toBeInTheDocument();
    expect(new Date(getStartDatePicker().value)).toEqual(mockMeetup.start);
    expect(new Date(getFinishDatePicker().value)).toEqual(mockMeetup.finish);
    expect(getLocationInput()).toHaveValue(mockMeetup.place);
    expect(getImage().src).toBe(mockMeetup.image?.url);
  });

  it('should accept user input', async () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    await editFields();

    expect(getSubjectInput()).toHaveValue(mockUpdatedMeetup.subject);
    expect(getExcerptInput()).toHaveValue(mockUpdatedMeetup.excerpt);
    expect(screen.getByText(mockUser2.fullName)).toBeInTheDocument();
    expect(new Date(getStartDatePicker().value)).toEqual(
      mockUpdatedMeetup.start,
    );
    expect(new Date(getFinishDatePicker().value)).toEqual(
      mockUpdatedMeetup.finish,
    );
    expect(getLocationInput()).toHaveValue(mockUpdatedMeetup.place);
    expect(getImagePreview()).toBeInTheDocument();
    expect(getImage().src).toBe(mockImageWithUrl2.url);
  });

  it('should validate form fields', async () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    expect(getSubmitBtn()).toBeDisabled();

    await act(() => {
      userEvent.clear(getSubjectInput());
      fireEvent.blur(getSubjectInput());
    });
    expect(
      screen.getByText('formFields.meetup.topic.errorText'),
    ).toBeInTheDocument();

    await act(() => {
      userEvent.clear(getExcerptInput());
      fireEvent.blur(getExcerptInput());
    });
    expect(
      screen.getByText('formFields.meetup.description.errorText'),
    ).toBeInTheDocument();
  });

  it('should handle form submit', async () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    await editFields();

    await act(() => {
      userEvent.click(getSubmitBtn());
    });

    expect(mockMeetupUpdate).toHaveBeenCalledWith(mockUpdatedMeetup);
    expect(screen.getByText('editMeetupPage.title')).toBeInTheDocument();
  });

  it('should be able to navigate back to news page', async () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.click(getCancelBtn());
    });

    expect(screen.getByText('Meetups page')).toBeInTheDocument();
  });

  it('should be able to redirect to preview page', async () => {
    render(<EditMeetupPage />, { wrapper: MockRouter });

    await act(() => {
      userEvent.click(getPreviewBtn());
    });

    expect(screen.getByText('View meetup')).toBeInTheDocument();
  });

  it('should render a Loading spinner if meetup is undefined', () => {
    mockUseMeetup.mockReturnValue(undefined);
    render(<EditMeetupPage />, { wrapper: MockRouter });
    expect(screen.getByText('loadingText.meetup')).toBeInTheDocument();
  });

  it('should render a Loading spinner while meetup is loading', () => {
    const { meetupStore } = new RootStore();
    const mockLoadingMeetup = new Meetup(mockMeetup, meetupStore);
    mockLoadingMeetup.isLoading = true;
    mockUseMeetup.mockReturnValue(mockLoadingMeetup);

    render(<EditMeetupPage />, { wrapper: MockRouter });

    expect(screen.getByText('loadingText.meetup')).toBeInTheDocument();
  });

  it('should render Not Found page if an error occurred while loading the meetup', () => {
    const { meetupStore } = new RootStore();
    const mockFailedMeetup = new Meetup(mockMeetup, meetupStore);
    mockFailedMeetup.isInitialized = true;
    mockFailedMeetup.isError = true;
    mockUseMeetup.mockReturnValue(mockFailedMeetup);

    render(<EditMeetupPage />, { wrapper: MockRouter });

    expect(screen.getByText('notFoundPage.title')).toBeInTheDocument();
  });
});
