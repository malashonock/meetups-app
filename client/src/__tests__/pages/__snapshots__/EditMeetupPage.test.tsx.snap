// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditMeetupPage should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <section
      class="container"
    >
      <h1
        class="typography h1 heading"
      >
        editMeetupPage.title
      </h1>
      <div
        class="contentWrapper"
      >
        <div
          class="textSection main"
        >
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="image"
            >
              formFields.meetup.image.label
            </label>
            <div
              class="preview large"
              data-testid="image-preview"
            >
              <figure
                class="image"
              >
                <img
                  alt="imagePreview.imgAlt"
                  src="http://localhost/test.jpg"
                />
              </figure>
              <button
                class="clearBtn"
                data-testid="clear-button"
                type="button"
              >
                <svg>
                  change-photo.svg
                </svg>
              </button>
            </div>
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="subject"
            >
              formFields.meetup.topic.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="subject"
              name="subject"
              type="text"
              value="Test meetup topic"
            />
          </div>
          <div
            class="dates"
          >
            <div
              class="container"
            >
              <label
                class="inputLabel label"
                for="start"
              >
                formFields.meetup.datetimeStart.label
              </label>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <span
                    aria-live="polite"
                    class="react-datepicker__aria-live"
                    role="alert"
                  />
                  <input
                    autocomplete="off"
                    class="input default"
                    id="start"
                    name="start"
                    placeholder="dateTimePicker.defaultPlaceholder"
                    type="text"
                    value="15 Mar 2023 14:00"
                  />
                </div>
              </div>
            </div>
            <div
              class="container"
            >
              <label
                class="inputLabel label"
                for="finish"
              >
                formFields.meetup.datetimeFinish.label
              </label>
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <span
                    aria-live="polite"
                    class="react-datepicker__aria-live"
                    role="alert"
                  />
                  <input
                    autocomplete="off"
                    class="input default"
                    id="finish"
                    name="finish"
                    placeholder="dateTimePicker.defaultPlaceholder"
                    type="text"
                    value="15 Mar 2023 15:30"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="place"
            >
              formFields.meetup.location.label
            </label>
            <input
              class="input default"
              data-testid="text-input"
              id="place"
              name="place"
              type="text"
              value="room 123"
            />
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="speakers"
            >
              formFields.meetup.speakers.label
            </label>
            <div
              data-testid="select-field"
            >
              <div
                class=" css-b62m3t-container"
                id="speakers"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="input default css-13cymwt-control"
                >
                  <div
                    class=" css-3w2yfm-ValueContainer"
                  >
                    <div
                      class=" css-1p3m7a8-multiValue"
                    >
                      <div
                        class=" css-wsp0cs-MultiValueGeneric"
                      >
                        John Doe
                      </div>
                      <div
                        aria-label="Remove John Doe"
                        class=" css-12a83d4-MultiValueRemove"
                        role="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="css-tj5bde-Svg"
                          focusable="false"
                          height="14"
                          viewBox="0 0 20 20"
                          width="14"
                        >
                          <path
                            d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                          />
                        </svg>
                      </div>
                    </div>
                    <div
                      class=" css-qbdosj-Input"
                      data-value=""
                    >
                      <input
                        aria-autocomplete="list"
                        aria-expanded="false"
                        aria-haspopup="true"
                        aria-label="speakers"
                        autocapitalize="none"
                        autocomplete="off"
                        autocorrect="off"
                        class=""
                        id="react-select-2-input"
                        role="combobox"
                        spellcheck="false"
                        style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                        tabindex="0"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <div
                      aria-hidden="true"
                      class=" css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
                        />
                      </svg>
                    </div>
                    <span
                      class=" css-1u9des2-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="speakers"
                    type="hidden"
                    value="[object Object]"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="container"
          >
            <label
              class="inputLabel label"
              for="excerpt"
            >
              formFields.meetup.description.label
            </label>
            <div
              class="container"
            >
              <textarea
                class="input default textArea"
                data-testid="text-area"
                id="excerpt"
                name="excerpt"
                rows="1"
                style="height: 2px;"
              >
                Test meetup description
              </textarea>
            </div>
          </div>
        </div>
        <div
          class="textSection actions"
        >
          <button
            class="actionButton cancelBtn button default"
            id="btn-cancel"
            type="button"
          >
            formButtons.cancel
          </button>
          <div
            class="actionsWrapper"
          >
            <button
              class="actionButton previewBtn button secondary"
              id="btn-preview"
              type="button"
            >
              formButtons.preview
            </button>
            <button
              class="actionButton saveBtn button primary"
              disabled=""
              id="btn-save"
              type="submit"
            >
              formButtons.save
            </button>
          </div>
        </div>
      </div>
    </section>
  </form>
</DocumentFragment>
`;
